src/wallet/api/wallet2_api.h

```diff
@@ -1144,6 +1210,330 @@ struct Wallet
+    virtual bool isFrozen(const std::string &key_image) const = 0;
+    /**
+    * brief: createOneOffSubaddress - create a subaddress for given index
+    * param: account_index - major index
+    * param: address_index - minor index
+    */
+    virtual void createOneOffSubaddress(std::uint32_t account_index, std::uint32_t address_index) = 0;
+    /**
+    * brief: getWalletState - get information about the wallet
+    * return: WalletState object
+    */
+    virtual WalletState getWalletState() const = 0;
+    /**
+    * brief: rewriteWalletFile - rewrite the wallet file for wallet update
+    * param: wallet_name - name of the wallet file (should exist)
+    * param: password - wallet password
+    * note: sets status error on fail
+    */
+    virtual void rewriteWalletFile(const std::string &wallet_name, const std::string &password) = 0;
+    /**
+    * brief: writeWatchOnlyWallet -  create a new watch-only wallet file with view keys from current wallet
+    * param: password - password for new watch-only wallet
+    * outparam: new_keys_file_name - wallet_name + "-watchonly.keys"
+    * note: sets status error on fail
+    */
+    virtual void writeWatchOnlyWallet(const std::string &password, std::string &new_keys_file_name) = 0;
+    /**
+    * brief: updatePoolState -
+    * outparam: process_txs - [ [tx, tx_id, double_spend_seen], ... ]
+    * param: refreshed - (default: false)
+    * param: try_incremental - (default: false)
+    * note: sets status error on fail
+    * This public method is typically called to make sure that the wallet's pool state is up-to-date by
+    * clients like simplewallet and the RPC daemon. Before incremental update this was the same method
+    * that 'refresh' also used, but now it's more complicated because for the time being we support
+    * the "old" and the "new" way of updating the pool and because only the 'getblocks' call supports
+    * incremental update but we don't want any blocks here.
+    *
+    * simplewallet does NOT update the pool info during automatic refresh to avoid disturbing interactive
+    * messages and prompts. When it finally calls this method here "to catch up" so to say we can't use
+    * incremental update anymore, because with that we might miss some txs altogether.
+    */
+    virtual void updatePoolState(std::vector<std::tuple<cryptonote::transaction, std::string, bool>> &process_txs, bool refreshed = false, bool try_incremental = false) = 0;
+    /**
+    * brief: processPoolState -
+    * param: txs - [ [tx, tx_id, double_spend_seen], ... ]
+    * note: sets status error on fail
+    */
+    virtual void processPoolState(const std::vector<std::tuple<cryptonote::transaction, std::string, bool>> &txs) = 0;
+    /**
+    * brief: getEnoteDetails - get information about all enotes
+    * outparam: enote_details -
+    */
+    virtual void getEnoteDetails(std::vector<std::shared_ptr<EnoteDetails>> &enote_details) const = 0;
```

Comment:
Should be `std::unique_ptr` instead of `std::shared_ptr`. We also need to include `<memory>` since we're using classes from that header now.